connector/connector_ldap.go

```diff
@@ -217,11 +238,74 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+type LDAPPool struct {
+	m          sync.Mutex
+	conns      map[*ldap.Conn]*ldap.Conn
+	initConn   int
+	maxConn    int
+	serverHost string
+	serverPort uint16
+	useTLS     bool
+	useSSL     bool
+	tlsConfig  *tls.Config
+}
+
+func (p *LDAPPool) fillPool() error {
+	if p.conns == nil {
+		p.conns = make(map[*ldap.Conn]*ldap.Conn)
+	}
+	for i := len(p.conns); i < p.initConn; i++ {
+		ldapConn, err := p.ldapConnect()
+		if err != nil {
+			return err
+		}
+		p.conns[ldapConn] = ldapConn
+	}
+	return nil
+}
+
+func (p *LDAPPool) Get() (*ldap.Conn, error) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	if p.conns == nil {
+		err := p.fillPool()
+		if err != nil {
+			return nil, err
+		}
+	}
+	for conn := range p.conns {
+		delete(p.conns, conn)
+		return conn, nil
+	}
+	// the pool is empty
+	conn, err := p.ldapConnect()
```

Comment:
Yes of course, thanks. One path could be to allocate and store a new ldap.Conn in the map while holding the mutex, release the mutex and then initialize/dial the connection. If that is not allowed by the library I will have to reverse it. Create+Dial the connection and then lock+store+unlock and return. Will fix it.